### PR TITLE
MA-4012: Updated kms.js and lambda.js lib files with additional funct…

### PR DESCRIPTION
**Pull request by bhargavapratik**
MA-4012: Updated kms.js and lambda.js lib files with additional functionalities .

### CODE REVIEW

File: aws_promise/kms.js [hunk 1]
```diff
@@ -20,5 +20,37 @@ module.exports = {
...
+	KeyId: input.keyId /* required */
+    };
+    return kms.listKeyPolicies(params).promise();
+  },
+
+  getkeyPolicy: function(input) {
+    var kms = this.findService(input);
+    var params = {
+	KeyId: input.keyId, /* required */ // The identifier of the CMK whose key policy you want to retrieve. You can use the key ID or the Amazon Resource Name (ARN) of the CMK.
+	PolicyName: input.policyName /* required */
```
Comment (mayankbharadwaj) [inline aws_promise/kms.js]: Can u add default policy name in case its not provided.
        PolicyName: "default"

File: aws_promise/lambda.js [hunk 2]
```diff
@@ -53,5 +53,21 @@ module.exports = {
       params['SourceArn'] = input.sourceArn;
     }
     return lambda.addPermission(params).promise();
+  },
+
+  invokeFunction: function(input) {
+    var lambda = this.findService(input);
+    var invocationType= (input.invocationType) ? input.invocationType : "RequestResponse";
+    var logType = (input.logType) ? input.logType : "Tail";
+    var qualifier= (input.qualifier) ? input.qualifier : "$LATEST";
```
Comment (mayankbharadwaj) [inline aws_promise/lambda.js:62]: add  below line to make payload string from object and pass this string in payload
    var payload = JSON.stringify(input.payload);

File: aws_promise/lambda.js [hunk 3]
```diff
@@ -53,5 +53,21 @@ module.exports = {
...
+    var params = {
+      FunctionName: input.functionArn,
+      ClientContext: input.clientContext, 
+      InvocationType: invocationType, 
+      LogType: logType, 
+      Payload: new Buffer(input.payload, 'base64'),
```
Comment (mayankbharadwaj) [inline aws_promise/lambda.js]: pass payload string as input in buffer object.

**Review by mayankbharadwaj — CHANGES_REQUESTED** [review]

File: aws_promise/lambda.js (hunk 2 above)
Comment (bhargavapratik) [reply aws_promise/lambda.js:62]: Done.

File: aws_promise/lambda.js (hunk 3 above)
Comment (bhargavapratik) [reply aws_promise/lambda.js]: Done.

File: aws_promise/kms.js [hunk 4]
```diff
@@ -20,5 +20,39 @@ module.exports = {
...
+  PolicyName: "default"// The name of the key policy.
+ }; */
+    return kms.getKeyPolicy(params).promise();
```
Comment (mayankbharadwaj) [inline aws_promise/kms.js]: make this as kms.putKeyPolicy(params).promise();

**Review by mayankbharadwaj — CHANGES_REQUESTED** [review]

File: aws_promise/lambda.js [hunk 5]
```diff
@@ -53,5 +53,22 @@ module.exports = {
...
+    var invocationType= (input.invocationType) ? input.invocationType : "RequestResponse";
+    var logType = (input.logType) ? input.logType : "Tail";
+    var qualifier= (input.qualifier) ? input.qualifier : "$LATEST";
+    var payload = JSON.stringify(input.payload);
+    var params = {
+      FunctionName: input.functionArn,
+      ClientContext: input.clientContext, 
+      InvocationType: invocationType, 
+      LogType: logType, 
+      Payload: new Buffer(payload, 'base64'),
```
Comment (mayankbharadwaj) [inline aws_promise/lambda.js]: pass payload as string only facing issues while passing it as buffer.

File: aws_promise/kms.js (hunk 4 above)
Comment (bhargavapratik) [reply aws_promise/kms.js]: Done.

**Review by kameshsg — COMMENTED** [review]
LGTM

File: aws_promise/lambda.js (hunk 5 above)
Comment (bhargavapratik) [reply aws_promise/lambda.js]: Done.

**Review by kameshsg — APPROVED** [review]
LGTM

File: aws_promise/kms.js [hunk 6]
```diff
@@ -20,5 +20,39 @@ module.exports = {
...
+    var kms = this.findService(input);
+    var params = {
+	KeyId: input.keyId /* required */
+    };
+    return kms.listKeyPolicies(params).promise();
+  },
+
+  getkeyPolicy: function(input) {
+    var kms = this.findService(input);
+    var policyName = (input.policyName) ?  input.policyName : "default";
```
Comment (SonalOjha) [inline aws_promise/kms.js:35]: what is default policy here?

File: aws_promise/kms.js [hunk 7]
```diff
@@ -20,5 +20,39 @@ module.exports = {
...
+
+  putkeyPolicy: function(input) {
+    var kms = this.findService(input);
+    var policyName = (input.policyName) ?  input.policyName : "default";
+    var params = {
+	KeyId: input.keyId, /* required */
+	Policy: input.policy, /* required */
+	PolicyName: policyName /* required */
+    };
+  /* var params = {
```
Comment (SonalOjha) [inline aws_promise/kms.js:51]: Is the code linted?

File: aws_promise/lambda.js [hunk 8]
```diff
@@ -53,5 +53,23 @@ module.exports = {
       params['SourceArn'] = input.sourceArn;
     }
     return lambda.addPermission(params).promise();
+  },
+
+  invokeFunction: function(input) {
+    var lambda = this.findService(input);
+    var invocationType= (input.invocationType) ? input.invocationType : "RequestResponse";
```
Comment (SonalOjha) [inline aws_promise/lambda.js:60]: Is the code linted?

File: aws_promise/lambda.js [hunk 9]
```diff
@@ -53,5 +53,23 @@ module.exports = {
...
+    var logType = (input.logType) ? input.logType : "Tail";
+    var qualifier= (input.qualifier) ? input.qualifier : "$LATEST";
+    var payload = JSON.stringify(input.payload);
+    var params = {
+      FunctionName: input.functionArn,
+      ClientContext: input.clientContext, 
+      InvocationType: invocationType, 
+      LogType: logType, 
+//      Payload: new Buffer(payload, 'base64'),
```
Comment (SonalOjha) [inline aws_promise/lambda.js]: May want to remove this line

File: aws_promise/s3bucket.js [hunk 10]
```diff
@@ -76,6 +76,106 @@ module.exports = {
       params.ACL = input.acl;
     }
     return s3.putObject(params).promise();
-  }
+  },
+
+  addBucketPolicy: function(input) {
+    var s3 = this.findService(input);
+    var policy = {
+        "Id": "PolicyID-"+input.uid+"",
```
Comment (SonalOjha) [inline aws_promise/s3bucket.js:84]: Can we have the static part of the policy in a configuration file

File: aws_promise/s3bucket.js [hunk 11]
```diff
@@ -76,6 +76,106 @@ module.exports = {
...
+                            FilterRules: [
+                                { Name: 'Prefix', Value: input.prefix},
+                                { Name: 'Suffix', Value: input.suffix}
+                            ]
+                        }
+                    },
+                    Id: input.uid
+                }
+        ];
+	params.NotificationConfiguration = {LambdaFunctionConfigurations: lambdaFunctionConfigurations};
```
Comment (SonalOjha) [inline aws_promise/s3bucket.js:149]: lint the code

File: aws_promise/kms.js (hunk 6 above)
Comment (bhargavapratik) [reply aws_promise/kms.js:35]: It is Policy Name , not policy.

File: aws_promise/kms.js (hunk 7 above)
Comment (bhargavapratik) [reply aws_promise/kms.js:51]: No, as this is old file.

Comment (SonalOjha) [reply aws_promise/kms.js:51]: Maybe you want to lint the code what being added.

**Review by SonalOjha — CHANGES_REQUESTED** [review]
Kindly lint the code which is being added to the file

File: aws_promise/lambda.js (hunk 8 above)
Comment (bhargavapratik) [reply aws_promise/lambda.js:60]: No, as this is old file.

File: aws_promise/s3bucket.js (hunk 10 above)
Comment (bhargavapratik) [reply aws_promise/s3bucket.js:84]: This is same as we have in BitBucket, and it's optional.

File: aws_promise/s3bucket.js (hunk 11 above)
Comment (bhargavapratik) [reply aws_promise/s3bucket.js:149]: It is old file, and hence not linted code.

File: aws_promise/kms.js (hunk 6 above)
Comment (RajkumarShivage) [reply aws_promise/kms.js:35]: What are the permissions associated with default policy?

**Review by SonalOjha — APPROVED** [review]
LGTM

**Review by mayankbharadwaj — APPROVED** [review]
LGTM

**Pull request merged**